ext/telemetry.c

```diff
@@ -0,0 +1,96 @@
+        return;
+    }
+
+    uint8_t formatted_run_time_id[36];
+    ddtrace_format_runtime_id(formatted_run_time_id);
+    ddog_CharSlice runtime_id = (ddog_CharSlice){ .ptr = (char *)formatted_run_time_id, .len = sizeof(formatted_run_time_id) };
+    dd_telemetry_instance_id = ddog_sidecar_instanceId_build(runtime_id, runtime_id);
+    char *agent_url = ddtrace_agent_url();
+    ddog_sidecar_session_config_setAgentUrl(&dd_sidecar, runtime_id, (ddog_CharSlice){ .ptr = agent_url, .len = strlen(agent_url) });
+    free(agent_url);
+
+    dd_composer_hook_id = zai_hook_install(ZAI_STRING_EMPTY, ZAI_STRING_EMPTY, dd_check_for_composer_autoloader, NULL, ZAI_HOOK_AUX_UNUSED, 0);
+}
+
+void ddtrace_telemetry_shutdown(void) {
+    if (dd_telemetry_instance_id) {
+        ddog_sidecar_instanceId_drop(dd_telemetry_instance_id);
+        ddog_sidecar_transport_drop(dd_sidecar);
+    }
+}
+
+void ddtrace_telemetry_finalize(void) {
+    if (!dd_telemetry_instance_id) {
+        return;
+    }
+
+    for (uint8_t i = 0; i < zai_config_memoized_entries_count; i++) {
+        zai_config_memoized_entry *cfg = &zai_config_memoized_entries[i];
+        zend_ini_entry *ini = cfg->ini_entries[0];
+#if ZTS
+        ini = zend_hash_find_ptr(EG(ini_directives), ini->name);
+#endif
+        if (!zend_string_equals_cstr(ini->value, cfg->default_encoded_value.ptr, cfg->default_encoded_value.len)) {
+            if (!zend_string_equals_literal(ini->name, "datadog.trace.enabled")) { // datadog.trace.enabled is meaningless: always off at rshutdown
+                ddog_sidecar_telemetry_enqueueConfig(&dd_sidecar, dd_telemetry_instance_id, &DDTRACE_G(telemetry_queue_id),
+                                                     dd_zend_string_to_CharSlice(ini->name), dd_zend_string_to_CharSlice(ini->value),
+                                                     cfg->name_index >= 0 ? DDOG_CONFIGURATION_ORIGIN_ENV_VAR : DDOG_CONFIGURATION_ORIGIN_CODE);
+            }
+        }
+    }
+
+    ddog_CharSlice service_name = DDOG_CHARSLICE_C("unnamed-php-service");
+    if (DDTRACE_G(last_flushed_root_service_name)) {
+        service_name = dd_zend_string_to_CharSlice(DDTRACE_G(last_flushed_root_service_name));
+    }
+
+    ddog_CharSlice php_version = dd_zend_string_to_CharSlice(Z_STR_P(zend_get_constant_str(ZEND_STRL("PHP_VERSION"))));
+    struct ddog_RuntimeMeta *meta = ddog_sidecar_runtimeMeta_build(DDOG_CHARSLICE_C("php"), php_version, DDOG_CHARSLICE_C(PHP_DDTRACE_VERSION));
+
+    ddog_sidecar_telemetry_flushServiceData(&dd_sidecar, dd_telemetry_instance_id, &DDTRACE_G(telemetry_queue_id), meta, service_name);
+
+    ddog_sidecar_runtimeMeta_drop(meta);
+
+    ddog_sidecar_telemetry_end(&dd_sidecar, dd_telemetry_instance_id, &DDTRACE_G(telemetry_queue_id));
+}
+
+void ddtrace_telemetry_notify_integration(const char *name, size_t name_len) {
+    if (dd_telemetry_instance_id) {
+        ddog_CharSlice integration = (ddog_CharSlice) {.len = name_len, .ptr = name};
```

Comment:
Ntipicking but we don't specify if the integration has been disabled by the customer (trough config) or if we get errors in the integration? Cf the [telemetry object](https://github.com/DataDog/instrumentation-telemetry-api-docs/blob/main/GeneratedDocumentation/ApiDocs/v2/SchemaDocumentation/Schemas/integration.md)